### PR TITLE
Directly send tensor via jit serialization

### PR DESCRIPTION
**Pull request by ZiyueXu77**
Fixes # .

### Description

Directly send tensor without converting to numpy
Using jit serialization to avoid pickle

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: examples/advanced/llm_hf/src/hf_sft_peft_fl.py [hunk 1]
```diff
@@ -232,8 +238,10 @@ def evaluate(input_weights, mode):
             for key in list(out_param.keys()):
                 out_param["model." + key] = out_param.pop(key).cpu()
 
-        # cast out_param to float32 preparing for communication
-        out_param = {k: v.to(torch.float32) for k, v in out_param.items()}
+        if args.message_mode.lower() == "numpy":
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/src/hf_sft_peft_fl.py:241]: If we want to use the numpy to_nvflare_converter (in case some filters only work with numpy)

File: examples/advanced/llm_hf/sft_job.py [hunk 2]
```diff
@@ -87,11 +89,26 @@ def main():
         site_name = f"site-{client_id}"
         data_path_train = os.path.join(args.data_path, client_id, "training.jsonl")
         data_path_valid = os.path.join(args.data_path, client_id, "validation.jsonl")
-        runner = ScriptRunner(
-            script=train_script,
-            script_args=f"--model_name_or_path {model_name_or_path} --data_path_train {data_path_train} --data_path_valid {data_path_valid} --output_path {output_path} --train_mode {train_mode} --clean_up {clean_up}",
-        )
+
+        if message_mode == "tensor":
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/sft_job.py:93]: may have some risks at this point, since most of our filters (DP, HE, etc.) works with numpy, need to test their compatibility with tensor

File: examples/advanced/llm_hf/README.md [hunk 3]
```diff
@@ -125,6 +125,13 @@ For message reduce, from float32 to 16-/8-/4-bit, the message size (in MB) of Ll
 
 Note that quantization will generate additional meta data, which can be significant for 4-bit cases.
 
+## Model Communication with Tensor
+In addition, since the model is trained with bf16, instead of first converting to numpy in float32, we can directly communicate with tensor in bf16 to avoid the message size inflation due to the conversion.
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/README.md:129]: mixed precision is common, presumebly yes - if no conflict with desired filters, we can always enable direct tensor communication

**Comment by ZiyueXu77** [comment]
> > > the logics seems only for LLM BF16. I think what want to achieve is for all Tensor, regardless or not.
> > 
> > 
> > I think we mixed two processes: conversion for filtering, and conversion for communication
> > Considering local to server communication (reverse will be similar) with quantization: local model --> to_nvflare_converter --> quant_filter --> (decomposer) --> communication --> (composer) --> dequant_filter --> global
> > Currently our client api executor has a default to_nvflare_converter as PTtoNumpy, so afterwards everything will be in numpy, including the serialization part, so the tensor decomposer/composer will not be called.
> > Now if instead of PTtoNumpy, we use a simple "pass through" to_nvflare_converter, it gonna have two indications:
> > 
> > 1. filter needs to handle tensor properly
> > 2. decomposer and composer will be needed to handle tensor communication, and currently this is again via numpy.
> > 
> > Hence we can have two places with tensor<->numpy conversion: converter for filter, and decomposer for communication. The first will mean all the following computations (filter) are in numpy, while the second means only the communication/serialization is via numpy - but it will be recovered to tensor once received, so "virtually" the whole pipeline is still in tensor.
> > For the sake of serialization efficency, my guess is that numpy maybe more efficient than jit (@nvidianz to confirm), then jit is only needed for formats not supported by numpy (e.g. bf16), but if otherwise, we can use jit for all cases (And maybe "safe tensor" as suggested).
> 
> The reason for avoid to_numpy() conversion is to avoid loss the Tensor Compression ratio to make sure the Tensor Model doesn't increase after transfer. It doesn't matter this conversion is in filter or other places, if we convert tensor in jit in place, and use to_n umpy in another place before sending over the wire, we already loss the compression, it JIT conversion is becomes pointless.
> 
> We need Tensor native in all communication pipeline

no this is not the case, use numpy + jit for serialization will not lead to bigger message, only the conversion for filter purpose will - because for that we want everything in numpy

File: nvflare/app_opt/pt/decomposers.py [hunk 4]
```diff
@@ -22,18 +22,63 @@
...
+        super().__init__()
+        self.register_buffer("saved_tensor", tensor)
+
+
 class TensorDecomposer(fobs.Decomposer):
     def supported_type(self):
         return torch.Tensor
 
     def decompose(self, target: torch.Tensor, manager: DatumManager = None) -> Any:
+        if target.dtype == torch.bfloat16:
```
[the remark this answers is not in the document]
Comment (nvidianz) [reply nvflare/app_opt/pt/decomposers.py:36]: JIT is twice slower.

**Review by nvidianz — DISMISSED** [review]

**Review by nvidianz — DISMISSED** [review]

File: examples/advanced/llm_hf/sft_job.py [hunk 5]
```diff
@@ -87,11 +89,26 @@ def main():
...
+            job.to(PTSendParamsConverter(), site_name, id="pt_send")
+            job.to(PTReceiveParamsConverter(), site_name, id="pt_receive")
+            runner = BaseScriptRunner(
+                script=train_script,
+                script_args=f"--model_name_or_path {model_name_or_path} --data_path_train {data_path_train} --data_path_valid {data_path_valid} --output_path {output_path} --train_mode {train_mode} --message_mode {message_mode} --clean_up {clean_up}",
+                from_nvflare_converter_id="pt_receive",
+                to_nvflare_converter_id="pt_send",
+            )
+        elif message_mode == "numpy":
+            runner = BaseScriptRunner(
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/sft_job.py:104]: that's the point, otherwise needs some deeper updates to our BaseScriptRunner, which is irrelevant to this PR

[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/sft_job.py:105]: updated

File: nvflare/app_common/executors/in_process_client_api_executor.py [hunk 6]
```diff
@@ -200,12 +200,11 @@ def send_data_to_peer(self, shareable, fl_ctx: FLContext):
 
     def _init_converter(self, fl_ctx: FLContext):
         engine = fl_ctx.get_engine()
-        from_nvflare_converter: ParamsConverter = engine.get_component(self._from_nvflare_converter_id)
+        from_nvflare_converter: ParamsConverter = engine.get_component(self._from_nvflare_converter_id[0])
         if from_nvflare_converter is not None:
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply nvflare/app_common/executors/in_process_client_api_executor.py:204]: corrected

File: nvflare/app_opt/quantization/dequantizor.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+from typing import Union
+
+import numpy as np
+import torch
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply nvflare/app_opt/quantization/dequantizor.py]: moved

File: nvflare/app_opt/quantization/quantizor.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+from typing import Union
+
+import numpy as np
+import torch
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply nvflare/app_opt/quantization/quantizor.py]: moved
